File: tests/python/cpp_src/test_stl.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <tvm/ffi/extra/stl.h>
#include <tvm/ffi/function.h>

#include <array>
#include <functional>
#include <map>
#include <numeric>
#include <optional>
#include <tuple>
#include <unordered_map>
#include <variant>
#include <vector>

namespace {

auto test_tuple(std::tuple<int, float> arg) -> std::tuple<float, int> {
  return std::make_tuple(std::get<1>(arg), std::get<0>(arg));
}

auto test_vector(std::optional<std::vector<std::array<int, 2>>> arg)
    -> std::optional<std::vector<int>> {
  if (arg) {
    auto result = std::vector<int>{};
    result.reserve(arg->size());
    for (const auto& row : *arg) {
      result.push_back(std::accumulate(row.begin(), row.end(), 0));
    }
    return result;
  } else {
    return std::nullopt;
  }
}

auto test_variant(std::variant<int, float, std::vector<std::variant<int, float>>> arg)
    -> std::variant<std::string, std::vector<std::string>> {
  if (std::holds_alternative<int>(arg)) {
    return "int";
  } else if (std::holds_alternative<float>(arg)) {
    return "float";
  } else {
    auto result = std::vector<std::string>{};
    for (const auto& item : std::get<std::vector<std::variant<int, float>>>(arg)) {
      if (std::holds_alternative<int>(item)) {
        result.emplace_back("int");
      } else {
        result.emplace_back("float");
      }
    }
    return result;
  }
}

auto test_map(const std::map<std::string, int>& map) -> std::map<int, std::string> {
  auto result = std::map<int, std::string>{};
  for (const auto& [key, value] : map) {
    result[value] = key;
  }
  return result;
}

auto test_map_2(const std::unordered_map<std::string, int>& map)
    -> std::unordered_map<int, std::string> {
  auto result = std::unordered_map<int, std::string>{};
  for (const auto& [key, value] : map) {
    result[value] = key;
  }
  return result;
}

auto test_function(std::function<int(int)> f) -> std::function<int(int)> {
  return [fn = std::move(f)](int value) { return fn(value) + 1; };
}

TVM_FFI_DLL_EXPORT_TYPED_FUNC(test_tuple, test_tuple)
TVM_FFI_DLL_EXPORT_TYPED_FUNC(test_vector, test_vector)
TVM_FFI_DLL_EXPORT_TYPED_FUNC(test_variant, test_variant)
TVM_FFI_DLL_EXPORT_TYPED_FUNC(test_map, test_map)
TVM_FFI_DLL_EXPORT_TYPED_FUNC(test_map_2, test_map_2)
TVM_FFI_DLL_EXPORT_TYPED_FUNC(test_function, test_function)

}  // namespace

File: tests/python/test_stl.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
import pathlib

import pytest
import tvm_ffi.cpp
from tvm_ffi.core import TypeSchema
from tvm_ffi.module import Module


def test_stl() -> None:
    cpp_path = pathlib.Path(__file__).parent.resolve() / "cpp_src" / "test_stl.cc"
    output_lib_path = tvm_ffi.cpp.build(
        name="test_stl",
        sources=[str(cpp_path)],
        extra_cflags=["-DTVM_FFI_DLL_EXPORT_INCLUDE_METADATA=1"],
    )

    mod: Module = tvm_ffi.load_module(output_lib_path)

    assert list(mod.test_tuple([1, 2.5])) == [2.5, 1]
    assert mod.test_vector(None) is None
    assert list(mod.test_vector([[1, 2], [3, 4]])) == [3, 7]
    assert mod.test_variant(1) == "int"
    assert mod.test_variant(1.0) == "float"
    assert list(mod.test_variant([1, 1.0])) == ["int", "float"]
    assert dict(mod.test_map({"a": 1, "b": 2})) == {1: "a", 2: "b"}
    assert dict(mod.test_map_2({"a": 1, "b": 2})) == {1: "a", 2: "b"}
    assert mod.test_function(lambda value: value)(10) == 11

    metadata = mod.get_function_metadata("test_function")
    assert metadata is not None
    schema = TypeSchema.from_json_str(metadata["type_schema"])
    assert schema.repr() == "Callable[[Callable[[int], int]], Callable[[int], int]]"

    vector_metadata = mod.get_function_metadata("test_vector")
    assert vector_metadata is not None
    vector_schema = TypeSchema.from_json_str(vector_metadata["type_schema"])
    assert vector_schema.repr() == "Callable[[Array[Array[int]] | None], Array[int] | None]"

    with pytest.raises(TypeError):
        mod.test_tuple([1.5, 2.5])
    with pytest.raises(TypeError):
        mod.test_vector([[1]])
    with pytest.raises(TypeError):
        mod.test_function(lambda value: value)()


if __name__ == "__main__":
    pytest.main([__file__])
